src/ipfs/daemon.rs

```diff
@@ -0,0 +1,221 @@
+    pub port:u16
+}
+
+pub struct IpfsDaemon {
+    conn_info: IpfsConnInfo,
+    client: IpfsClient,
+    tokio: Runtime
+}
+
+impl IpfsDaemon {
+    /// This method launches the IPFS daemon on the port/address given when the the instance was
+    /// created. During this proccess it also creates a thread that listens for a shutdown signal
+    /// and stops the IPFS daemon gracefully if the signal is given.
+    pub async fn launch(&self) -> Result<()>{
+        //launch the daemon
+        let api_addr = format!("/ip4/{}/tcp/{}", self.conn_info.address, self.conn_info.port);
+        println!("Launhing IPFS...");
+        Command::new(IPFS_EXE)
+            .arg("--api")
+            .arg(&api_addr)
+            .arg("daemon")
+            .spawn()
+            .unwrap_or_else(|e| panic!("Failed to start IPFS daemon: {}\n This error may be because the Kubo binary is not on your PATH.", e));
+
+        //Wait ipfs to ready
+        println!("Waiting for IPFS to ready..");
+        self.await_ready().await?;
+
+        //Reduce log level for IPFS
+        self.tokio.block_on(async {
+            self.client.log_level(Logger::All, LoggingLevel::Error).await
+        })?;
+        
+        //setup gracefull shutdown
+        println!("Creating gracefull shutdown for IPFS...");
+        let me = self.clone();
+        thread::spawn(move || {
+            let signal_guard = SignalGuard::new();
+
+            signal_guard.at_exit(move |sig| {
+                println!("Signal {} received. Attempting to stop IPFS...", sig);
+                match me.shutdown() {
+                    Ok(_) => println!("{}", "IPFS has shutdown successfully.".green()),
+                    Err(_) => println!("{}", "IPFS failed to shutdown succefully! You may need to stop the proccess yourself.".red())
+                };
+            });
+        });
+
+        println!("{}", "IPFS has launched successfully!!".green());
+        Ok(())
+    }
+
+    /// This method sends an http signal to the IPFS deamon to shutdown. 
+    /// 
+    /// Note 1: This method should work even if the daemon was not started by this proccess.
+    /// 
+    /// Note 2: This method return when it recieves an http response from the daemon, and not
+    /// when the proccess has actually stopped.
+    pub fn shutdown(&self) -> Result<()>{
+        self.tokio.block_on(async {
+            block_on(self.client.shutdown())
+        })?;
+        Ok(())
+    }
+
+    async fn is_ipfs_ready(&self) -> bool {
+        let res = self.tokio.block_on(async {
+            self.client.config_show().await
+        });
+        // if res.is_ok() {
+        //     println!("Config is {}", res.as_ref().unwrap());
+        // }
+        return match res {
+            Ok(_) => true,
+            Err(_) => false
+        };
+    }
+
+    async fn await_ready(&self) -> Result<()> {
+        let start_time = SystemTime::now();
+        loop {
+            println!("{}", "checking if ipfs is ready...".green());
```

Comment:
```suggestion
            println!("{}", "Checking if ipfs is ready...".green());
```

We should be consistent about capitalizing the first character of logging statements.